ast/check.go

```diff
@@ -146,7 +146,97 @@ func (tc *typeChecker) checkLanguageBuiltins(env *TypeEnv, builtins map[string]*
 	return env
 }
 
+func getPrefixToOverride(name string, env *TypeEnv, rule *Rule) (string, types.Type, *Error) {
+	ref, err := ParseRef(name)
+	if err != nil {
+		return "", nil, NewError(TypeErr, rule.Location, err.Error())
+	}
+	t := env.tree.Get(ref)
+	if t != nil {
+		return name, t, nil
+	}
+	names := strings.Split(name, ".")
+	if len(names) > 1 {
+		prefix, t, err := getPrefixToOverride(strings.TrimSuffix(name, "."+names[len(names)-1]), env, rule)
+		if err != nil {
+			return "", nil, err
+		}
+		if t != nil {
+			return prefix, t, nil
+		}
+	}
+	return "", nil, nil
+}
+
+func override(names []string, t types.Type, o types.Type) types.Type {
+	obj := t.(*types.Object)
+	staticProps := obj.StaticProperties()
+	newStaticProps := []*types.StaticProperty{}
+
+	for _, prop := range staticProps {
+		if len(names) == 1 && names[0] == prop.Key.(string) {
+			prop.Value = o
+		} else if len(names) > 0 && names[0] == prop.Key.(string) {
+			prop.Value = override(names[1:], prop.Value, o)
+		}
+		newStaticProps = append(newStaticProps, types.NewStaticProperty(prop.Key, prop.Value))
+	}
+	return types.NewObject(newStaticProps, nil)
+}
+
+// Annotations must immediately precede the rule definition and are of the form: #@rulesSchema=<expr>:<schema-key>
+func (tc *typeChecker) processAnnotation(annot *SchemaAnnotation, env *TypeEnv, rule *Rule) (Ref, types.Type, *Error) {
+	if env.schemaSet == nil || env.schemaSet.ByPath == nil {
+		return nil, nil, NewError(TypeErr, rule.Location, "Schemas need to be supplied for the annotation: %s", annot.Schema)
```

Comment:
💭  Do we always start with uppercase letters in errors...? If so, fine. If not, let's not do it here 😉

Reply:
Done

Reply:
Done